Fix columns width when wrapping text

rosen-vladimirov opened the pull request:
## The problem

Currently when table's width is above terminal width, all column widths will be recalculated and the final result is equal column widths. This leads to unwanted behavior for our use-case. for example:

| General | Usage |
| --- | --- |
| Android | some very long command with options and several other items |

In this case after wrapping the text, both columns will have the same width, which will lead to output, which will not look very good.
### Proposed solution

Use each column's required width and calculate its relative size based on current terminal width. For example if we have the following structure and columns width:

| Column A | Column B | Column C |
| --- | --- | --- |
| 30 | 50 | 20 |

and the available terminal width is 80, the calculation of the columns should be:
Column A = (30 / 100) \* 80 = 24
Column B = (50 / 100) \* 80 = 40
Column C = (20 / 100) \* 80  = 16
Column A + Column B + Column C = 80
### Extended solution

In most of our use-cases, we have 2 columns in our table - the first one contains short text (around 10 symbols) and the second column has some very long text. After applying suggested solution, first column's width is reduced and the text inside it is trimmed or placed on two rows. Add min_column_width property and if the width of the column is below this value, do not reduce columns size. In case the calculated value is below min_column_width, return min_column_width. The problem here is in the following scenario:

| Column A | Column B | Column C | Column D | Column E |
| --- | --- | --- | --- | --- |
| 10 | 20 | 30 | 10 | 30 |

The sum is 100. Max terminal width is 80.
Lets set the min_column_width property to 23. Here's the calculation:
Column A - 10 < 23 => 10
Column B - (20 / 100) \* 80 = 16
Column C - (30 / 100) \* 80 = 24
Column D - 10 < 23 => 10
Column E - (30 / 100) \* 80 = 24
A + B + C +D +E = 74
But some of the columns are less then already defined min_column_width(23). If we set them to min_column_width the values are:
Column A - 10 < 23 => 10
Column B - (20 / 100) \* 80 = 16 < 23 => 23
Column C - (30 / 100) \* 80 = 24
Column D - 10 < 23 => 10
Column E - (30 / 100) \* 80 = 24
A + B + C +D +E = 91

This way we've exceeded terminal width. The proposed solution is to modify the formula based on the remaining terminal and the remaining required size.
For each column we check if its size is below min_column_width. In case it is, we set this column as static (its size will not change) and we remove the width
from remaining terminal and remaining required width. If the width is above min_column_width we execute:
(required_col_width / remaining_required_size) \* remaining_terminal_width
In case the value is above min_column_width, we save it as dynamic size (could be changed later).
In case the value is below min_column_width, we save it as static size and remove it from remaining_terminal_width. The original value should be removed from remaining_required_size.
At this point we have to recalculate all dynamic values and apply the same logic as above.
## Implement own line splitting
### js-beautify issues

Remove usage of `js-beautify` module for spliting the string for multiline usage. There are two problems with `js-beautify` module:
- text that contains <...> is always split on several lines. `js-beautify` considers <...> as HTML tag and places is to new line.
- `js-beautify` does not guarantee that the string will be split based on wrap_line_length value. The string is split based on the first delimiter AFTER wrap_line_length.
### Proposed solution

Use own function instead of `js-beautify` to split the lines. The following rules are applied:
- if content is below wrap_line_length, return the content without modifying it
- if content is above wrap_line_length the rules are:
  - if the symbol on index "wrap_line_length" is delimiter, take all symbols up to it and split the rest of the text based on this rules (call the method with rest of the text)
  - if the next symbol is not delimiter, try to find the last delimiter in the symbols up to wrap_line_length index.
    - in case delimiter is found, split the line by it and split everything after it based on the same rules
    - in case delimiter is not found, split the word, this means - take all symbols up to wrap_line_length and split the rest of the string based on the same rules.
### 'Code' problem

If the passed text is colorized, the string contains several specific colorization symbols. They are rendered by the console as colors, but when checking string.length, they are calculated (typically the length of these symbols is 8-9). After we split the content on several lines, we lose the colorization of subsequent lines.
### 'Code' problem solution

When splitting the text, calculate the length of all "code" symbols which we can include until we reach the allowed length of the column. When splitting, check the number of "code" segments in the line - if they are odd number, get the last "code" segment and add it to the beginning of the remaining text. This way we have the correct colorization on the next lines.

rosen-vladimirov committed:
Fix columns width when wrapping text

Currently when table's width is above terminal width, all column widths will be recalculated and the final result is equal column widths. This leads to unwanted behavior for our use-case. for example:
General | Usage
Android | some very long command with options
In this case after wrapping the text, both columns will have the same width, which will lead to output like this:
General                          |                               Usage
Android                          |         some very long command with options

###Proposed solution
Use each column's required width and calculate its relative size based on current terminal width. For example if we have the following structure:
Column A | Column B | Column C
   30                  50                20
and the available terminal width is 80, the calculation of the columns should be:
Column A = (30 / 100) * 80 = 24
Column B = (50 / 100) * 80 = 40
Column C = (20 / 100) * 80  = 16
Column A + Column B + Column C = 80

###Extended solution
In most of our use-cases, we have 2 columns in our table - the first one contains short text (around 10 symbols) and the second column has some very long text. After applying suggested solution, first column's width is reduced and the text inside it is trimmed or placed on two rows. Add minColumnWidth property and if the width of the column is below this value, do not reduce columns size. In case the calculated value is below minColumnWidth, return minColumnWidth. The problem here is in the following scenario:
Column A | Column B | Column C | Column D | Column E
    10                   20                30                 10                 30
The sum is 100. Max terminal width is 80.
Lets set the minColumnWidth property to 23. Here's the calculation:
Column A - 10 < 25 => 10
Column B - (20 / 100) * 80 = 16
Column C - (30 / 100) * 80 = 24
Column D - 10 < 25 => 10
Column E - (30 / 100) * 80 = 24
A + B + C +D +E = 74
But some of the columns are less then already defined minColumnWidth(23). If we set them to minColumnWidth the values are:
Column A - 10 < 23 => 10
Column B - (20 / 100) * 80 = 16 < 23 => 23
Column C - (30 / 100) * 80 = 24
Column D - 10 < 23 => 10
Column E - (30 / 100) * 80 = 24
A + B + C +D +E = 91

This way we've exceeded terminal width. The proposed solution is to modify the formula based on the remaining terminal and the remaining required size.
For each column we check if its size is below minColumnWidth. In case it is, we set this column as static (its size will not change) and we remove the width
from remaining terminal and remaining required width. If the width is above minColumnWidth we execute:
(required_col_width / remaining_required_size) * remaining_terminal_width
In case the value is above minColumnWidth, we save it as dynamic size (could be changed later).
In case the value is below minColumnWidth, we save it as static size and remove it from remaining_terminal_width. The original value should be removed from remaining_required_size.
At this point we have to recalculate all dynamic values and apply the same logic as above.

ligaz commented on lib/index.js:
```diff
@@ -123,12 +124,37 @@ Table.prototype.toString = function (){
-        return new_col_width;
+    
+    if(terminal_width <= required_width){
+      var remainingTerminalWidth = terminal_width - (colWidths.length * 2);
```
Maybe it is good to follow the author convention for naming locals with underscores.

ligaz commented:
Adding a unit test will be great. :cop:

rosen-vladimirov commented:
I'll add unit tests and also I have to handle recalculating better as there's a problem in case when recalculated value is below minColumnWidth

rosen-vladimirov committed:
Add recalculateColWidths method

In cases when we recalculate column widths, we must be sure that new column width is at least min_column_width(for all columns, which are above min_column_width, after calculation we do not want to have value less than min_column_width - that's the point of this parameter).
Add warning in the case when calculated table width is above terminal width (it is possible when min_column_width has high value).

rosen-vladimirov commented:
Ok I've added recalculate_col_widths  method, I just have to write some unit test :)

teobugslayer commented on lib/index.js:
```diff
@@ -85,6 +86,7 @@ Table.prototype.toString = function (){
     , chars = options.chars
     , truncater = options.truncate
     , colWidths = options.colWidths || new Array(this.head.length)
+    , min_column_width = options.min_column_width || 15
```
Thy magic constant has imploded the universe! (Because, magic and stuff). 

On a more serious note, consider describing it, at least in a comment.

rosen-vladimirov replied:
Done. 
Added comment in code and description in README.md

rosen-vladimirov committed:
Update README

Update README with comments regarding new functionality. Add comments in index.js for min_column_width property.

rosen-vladimirov committed:
Split string correctly

Remove usage of `js-beautify` module for spliting the string for multiline usage. There are two problems with `js-beautify` module:
* text that contains <...> is always split on several lines. `js-beautify` considers <...> as HTML tag and places is to new line.
* `js-beautify` does not guarantee that the string will be split based on wrap_line_length value. The string is split based on the first delimiter AFTER wrap_line_length.

### Proposed solution
Use own function instead of `js-beautify` to split the lines. The following rules are applied:
* if content is below wrap_line_length, return the content without modifying it
* if content is above wrap_line_length the rules are:
    * if the symbol on index "wrap_line_length" is delimiter, take all symbols up to it and split the rest of the text based on this rules (call the method with rest of the text)
    * if the next symbol is not delimiter, try to find the last delimiter in the symbols up to wrap_line_length index.
        * in case delimiter is found, split the line by it and split everything after it based on the same rules
        * in case delimiter is not found, split the word, this means - take all symbols up to wrap_line_length and split the rest of the string based on the same rules.

### 'Code' problem
If the passed text is colorized, the string contains several specific colorization symbols. They are rendered by the console as colors, but when checking string.length, they are calculated (typically the length of these symbols is 8-9). After we split the content on several lines, we lose the colorization of subsequent lines.

### 'Code' problem solution
When splitting the text, calculate the length of all "code" symbols which we can include until we reach the allowed length of the column. When splitting, check the number of "code" segments in the line - if they are odd number, get the last "code" segment and add it to the beginning of the remaining text. This way we have the correct colorization on the next lines.

teobugslayer commented:
:+1:
